File: src/solar_history.py
'''
This script is used to get the history of the solar system.
It will get the history of the last 10 days.
It will also print the data to the console.

'''

import serial
import time
from datetime import datetime, timedelta

class SolarHistory:
    def __init__(self, port, baudrate=19200, timeout=1):
        self.port = port
        self.baudrate = baudrate
        self.timeout = timeout
        self.ser = None
        self.buffer = b''
        self.base_command = ''
        self.day_sequence = None
        
        # States
        (GET_DAY, WAIT_HEADER, WAIT_CR, FOUND) = range(4)
        self.GET_DAY = GET_DAY
        self.WAIT_HEADER = WAIT_HEADER
        self.WAIT_CR = WAIT_CR
        self.FOUND = FOUND
        self.state = self.GET_DAY

    def calculate_checksum(self, message):
        checksum = 0x55  # Initial value per VE.Direct protocol
        for byte in message:
            checksum -= byte
        checksum = checksum & 0xFF  # Ensure 8-bit result
        return bytes([checksum])

    def bytes_to_hex_string(self, byte_data):
        return ''.join([f'{b:02X}' for b in byte_data])

    def hex_to_bytes(self, hex_string):
        return bytes.fromhex(hex_string)

    def connect(self):
        try:
            self.ser = serial.Serial(self.port, self.baudrate, timeout=self.timeout)
            print(f"Connected to {self.port}")
            return True
        except Exception as e:
            print(f"Error opening serial port: {e}")
            return False

    def disconnect(self):
        if self.ser and self.ser.is_open:
            self.ser.close()

    def process_byte(self, byte):
        if not byte:
            return None

        self.buffer += byte

        if self.state == self.GET_DAY:
            # Build command for history day record
            self.base_command = '7' + str(50 + self.current_day) + '10' + '00'
            checksum = self.calculate_checksum(bytes.fromhex('0' + self.base_command))
            checksum_hex = self.bytes_to_hex_string(checksum)
            command = str.encode(':' + self.base_command + checksum_hex + '\n')
            self.ser.write(command)
            print('sent:', command)
            time.sleep(0.1)  # Wait for response
            self.state = self.WAIT_HEADER
            self.buffer = b''
            return None

        elif self.state == self.WAIT_HEADER:
            if self.buffer.endswith(self.base_command.encode()):
                self.state = self.WAIT_CR
                print('buffer:', self.buffer)
                self.buffer = b''

        elif self.state == self.WAIT_CR:
            if self.buffer.endswith(b'\n'):
                self.state = self.FOUND
                print('buffer:', self.buffer)
                response = self.buffer[64:68].decode('utf-8')
                response = bytes.fromhex(response)
                self.day_sequence = int.from_bytes(response, 'little')
                print('day_sequence:', self.day_sequence)
                return self.day_sequence

        elif self.state == self.FOUND:
            if len(self.buffer) >= 80:
                self.state = self.GET_DAY
                return self.day_sequence

        return None

    def get_history(self, days=10):
        if not self.connect():
            return

        try:
            self.current_day = 0
            while self.current_day < days:
                byte = self.ser.read(1)
                result = self.process_byte(byte)
                
                if result is not None:
                    print(f"Day {self.current_day} sequence: {result}")
                    self.current_day += 1
                    time.sleep(0.1)  # Small delay between requests
        except Exception as e:
            print(f"Error: {e}")
        finally:
            self.disconnect()

if __name__ == "__main__":
    # Example usage
    solar = SolarHistory('/dev/ttyUSB1')
    solar.get_history(10)

File: scripts_archive/solar_history.lua
-- Solar History Reader Script
-- This script reads solar history data from a VE.Direct device using the same protocol
-- as the Python solar_history.py script

---@diagnostic disable: param-type-mismatch
---@diagnostic disable: need-check-nil
---@diagnostic disable: cast-local-type

-- Configuration
local baud_rate = 19200
local days_to_read = 10

-- States for the state machine
local GET_DAY = 1
local WAIT_HEADER = 2
local WAIT_CR = 3
local FOUND = 4

-- Initialize serial port
local port = assert(serial:find_serial(0), "Could not find Scripting Serial Port 0")
print("Scripting Serial Port found")

-- Begin the serial port
port:begin(baud_rate)
port:set_flow_control(0)

-- State machine variables
local state = FOUND
local buffer = ""
local pbuffer = ""
local base_command = ""
local current_day = 0
local day_sequence = nil

-- Helper function to calculate checksum
local function calculate_checksum(message)
    local checksum = 0x55 -- Initial value per VE.Direct protocol
    for i = 1, #message do
        checksum = checksum - string.byte(message, i)
    end
    return checksum & 0xFF -- Ensure 8-bit result
end

-- Helper function to convert bytes to hex string
local function bytes_to_hex_string(bytes)
    local hex = ""
    for i = 1, #bytes do
        hex = hex .. string.format("%02X", string.byte(bytes, i))
    end
    return hex
end

-- Process a single byte through the state machine
local function process_byte(byte)
    if not byte then
        return nil
    end

    -- buffer = buffer .. string.char(byte)

    -- length of pbuffer
    -- print("pbuffer length: " .. #pbuffer)

    if state == GET_DAY then
        -- Build command for history day record
        -- base_command = "7" .. tostring(50 + current_day) .. "10" .. "00"
        -- local checksum = calculate_checksum("0" .. base_command)
        -- local checksum_hex = string.format("%02X", checksum)
        -- local command = ":" .. base_command .. checksum_hex .. "\n"


        local len_written = port:writestring(':7501000EE' .. '\n') -- send a command get todays history to the solar inverter
        -- print(string.format(":7501000EE, len_writ: %d", len_written))

        -- local len_written = port:writestring(command)
        -- print("len_written: " .. len_written)
        -- delay 100ms


        -- state = WAIT_HEADER
        -- print("state: WAIT_HEADER")
        state = FOUND
        print("state: FOUND")

        buffer = ""
        return nil
    elseif state == WAIT_HEADER then
        -- gcs:send_text(6, byte)
        if buffer:sub(- #base_command) == base_command then
            state = WAIT_CR
            print("state: WAIT_CR")
            buffer = ""
        end
    elseif state == WAIT_CR then
        if buffer:sub(-1) == "\n" then
            state = FOUND
            print("state: FOUND")

            -- Extract the sequence number from positions 64-68
            local response = buffer:sub(65, 68)
            day_sequence = tonumber(response, 16)
            print("day_sequence: " .. tostring(day_sequence))
            return day_sequence
        end
        -- elseif state == FOUND then
        --     if #buffer >= 80 then
        --         state = GET_DAY
        --         print("state: GET_DAY")
        --         return day_sequence
        --     end
    end

    return nil
end

-- Main update function that runs periodically
local count = 0
local function update()
    count = count + 1
    print("state: " .. state)
    print(string.format("count: %d", count))
    if count % 5 == 0 then
        state = GET_DAY
        print("state: GET_DAY")
    end

    local n_bytes = port:available()
    print(string.format("Bytes: %d", n_bytes:toint()))
    while n_bytes > 0 do
        -- only read a max of 512 bytes in a go
        -- this limits memory consumption
        -- buffer = {} -- table to buffer data
        local bytes_target = n_bytes - math.min(n_bytes, 512)
        while n_bytes > bytes_target do
            local byte = port:read()
            pbuffer = pbuffer .. string.char(byte)
            n_bytes = n_bytes - 1
            local result = process_byte(byte)
            -- if result then
            --     gcs:send_text(6, string.format("Day %d sequence: %d", current_day, result))
            --     current_day = current_day + 1

            --     if current_day >= days_to_read then
            --         print("Completed reading all days")
            --         return nil -- Stop the script
            --     end
            -- end
        end
        -- print("pbuffer: " .. pbuffer)
        pbuffer = ""
    end


    return update, 500 -- Reschedule the loop with 100ms delay
end

-- Start the script
print("Starting Solar History reader script")

return update, 1000 -- Run after 1000ms before starting to reschedule
